lib/project_types/extension/features/argo/base.rb

```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+require 'base64'
+require 'shopify_cli'
+require 'semantic/semantic'
+
+module Extension
+  module Features
+    module Argo
+      class Base
+        include SmartProperties
+
+        NPM_LIST_COMMAND = %w(list).freeze
+        YARN_LIST_COMMAND = %w(list --pattern).freeze
+        NPM_LIST_PARAMETERS = %w(--json --prod=true --depth=0).freeze
+        YARN_LIST_PARAMETERS = %w(--depth=0).freeze
+
+        SCRIPT_PATH = %w(build main.js).freeze
+
+        property! :setup, accepts: Features::ArgoSetup
+        property! :renderer_package, accepts: String
+
+        def create(directory_name, identifier, context)
+          setup.call(directory_name, identifier, context)
+        end
+
+        def config(context)
+          filepath = File.join(context.root, SCRIPT_PATH)
+          context.abort(context.message('features.argo.missing_file_error')) unless File.exist?(filepath)
+          begin
+            {
+              renderer_version: extract_argo_renderer_version(context),
+              serialized_script: Base64.strict_encode64(File.open(filepath).read.chomp),
+            }
+          rescue StandardError
+            context.abort(context.message('features.argo.script_prepare_error'))
+          end
+        end
+
+        private
+
+        def extract_argo_renderer_version(context)
+          renderer_package = self.renderer_package
+          js_system = ShopifyCli::JsSystem.new(ctx: context)
+          installed_package_manager = js_system.package_manager
+          result, error, stat = js_system.call(
+            yarn: YARN_LIST_COMMAND + [renderer_package] + YARN_LIST_PARAMETERS,
+            npm: NPM_LIST_COMMAND + [renderer_package] + NPM_LIST_PARAMETERS,
+            with_capture: true
+          )
```

Comment:
In a yarn managed environment, I use `yarn list` command to get the Argo renderer package info. Based on the tophatting results, it seems that yarn only use `package.json` as the source of truth. While I'm looking for a way to get the info from what already installed in the`node_module` folder to prevent any probable manual updates being pushed to Core. @vividviolet I was wondering if you could share your thoughts and/or knew a better way to get the package info.

Reply:
@sharmin-alvandi could you clarify what you mean by yarn only uses `package.json`? Yarn uses the `package.json` along with the `yarn.lock` file to manage installing dependencies and settings for compatible versions. However, the real version installed should be printed when you do `yarn list`. Having said that, I think something is off with the `yarn.lock` file - it seems to be referencing both `@shopify/argo-admin@0.4.1` and `@shopify/argo-admin@latest`. I'll have a look and get back to you.

Reply:
We'll update it to 0.4.2 on the app libs repo. Let you know!

Reply:
> @sharmin-alvandi could you clarify what you mean by yarn only uses `package.json`? Yarn uses the `package.json` along with the `yarn.lock` file to manage installing dependencies and settings for compatible versions. However, the real version installed should be printed when you do `yarn list`. Having said that, I think something is off with the `yarn.lock` file - it seems to be referencing both `@shopify/argo-admin@0.4.1` and `@shopify/argo-admin@latest`. I'll have a look and get back to you.

Thanks Trish!

Reply:
I gave this a quick look and it appears that we might be seeing multiple versions of `@shopify/argo-admin` because of the inter-dependency of **app-extension-libs** packages. Even though `argo-admin-template` is using `@shopify/argo-admin-host@latest`, `@shopify/argo-admin-host` itself is set to use `@shopify/argo-admin@0.4.1` (or whatever is the latest version when we execute `lerna release`)

_package.json_ from `@shopify/argo-admin-host`
<img width="575" alt="Screen Shot 2020-09-09 at 3 03 35 PM" src="https://user-images.githubusercontent.com/2709526/92642077-b2b9b380-f2ad-11ea-8e34-a689b0bedd22.png">

We could try updating packages in `app-extensions-libs` to `latest` and see if that would fix the problem? Not sure what the implications to that would be though, but I think it would be ok for that project since we're using the source directly in the playground, and unreleased changes wouldn't get picked up. But if there are cases where package versions needs to line up, it might be tricky.

<img width="1113" alt="Screen Shot 2020-09-09 at 2 51 15 PM" src="https://user-images.githubusercontent.com/2709526/92641900-6cfceb00-f2ad-11ea-86c9-48781b612751.png">

### Why did it work for `npm`?
We're probably able to get the latest version with npm because there's no `package-lock.json`, therefore we'll always get whatever is the latest.

Reply:
![image](https://user-images.githubusercontent.com/29458473/92646312-3bd3e900-f2b4-11ea-881c-8ab27e445e0f.png)

This screenshot indicates that it's picking up the wrong version, not that there's a version mismatch for the dependencies of `argo-admin-host`. `argo-admin-host` latest should have picked up `0.4.2` which will require the same version of `argo-admin` and `argo-admin-react`. The screenshot is showing it using `argo-admin-host@0.4.1`
`

Reply:
Ignore what I said above, it's irrelevant for the purpose of the PR. After pairing with Sharmin, what was happening is the behaviour of `npm list` is different than `yarn list`. With `yarn list`, it doesn't actually return the version number that's installed, it just displays whatever is inside `package.json`, even if it's not what's inside `node_modules`.

Apparently this is a known issue and others have ran into the same problem: https://github.com/yarnpkg/yarn/issues/3569 I'm going to write up a separate thing to explore other options, I think ultimately we need a reliable way to know which version of **@shopify/argo-admin** is being consumed by an extension, regardless if the project uses **npm** or **yarn**.

Reply:
@hannachen The issue linked seems to be related to dependency depths, not really that yarn doesn't list out the installed dependencies?

From my testing I'm getting the real installed version when doing `yarn list`. The package.json shows "latest" but the version coming back is `0.0.4.1`.

![argo-version](https://user-images.githubusercontent.com/29458473/92659739-fa006e00-f2c6-11ea-8be7-4d5061154c23.gif)

Reply:
@vividviolet What would be the result of `yarn list` if you change the 'latest' to an older version in package.json manually without running the yarn install afterward? In my testing the older version is returned which is not the actual installed version.

Reply:
I'm still getting the installed version:
![argo-version-older](https://user-images.githubusercontent.com/29458473/92661836-79903c00-f2cb-11ea-8219-05fd245c0fda.gif)

Make sure you're looking at the top level dependency, not dependencies within `argo-admin-cli`

Reply:
@vividviolet What if you do it without the `yarn install` step after updating package.json?

Reply:
Oh sorry, I see what you mean now - in this case I get the wrong version. This seems like a weird edge case though. If you change anything inside your package.json you need to do `yarn install` to install the dependencies. If you change the package without doing this it would it still points to the old modules. If another user checks out your project and does a yarn install they will get the new version (as expected).

Maybe we should enforce it to run `yarn install; yarn build` right before pushing the extension to Partner's.